config/pmd.xml

```diff
@@ -0,0 +1,42 @@
+<?xml version="1.0"?>
+
+<ruleset name="Basic"
+    xmlns="http://pmd.sourceforge.net/ruleset/2.0.0"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 https://pmd.sourceforge.io/ruleset_2_0_0.xsd">
+  <description>
+The Basic ruleset contains a collection of good practices which should be followed.
+  </description>
+
+    <!-- Rules, that have been moved into a category -->
+    <rule ref="category/java/errorprone.xml/AvoidBranchingStatementAsLastInLoop" deprecated="true" />
+    <rule ref="category/java/errorprone.xml/AvoidDecimalLiteralsInBigDecimalConstructor" deprecated="true" />
+    <rule ref="category/java/errorprone.xml/AvoidMultipleUnaryOperators" deprecated="true" />
+    <rule ref="category/java/errorprone.xml/AvoidUsingOctalValues" deprecated="true" />
+    <rule ref="category/java/errorprone.xml/BrokenNullCheck" deprecated="true" />
+    <rule ref="category/java/errorprone.xml/CheckSkipResult" deprecated="true" />
+    <rule ref="category/java/errorprone.xml/ClassCastExceptionWithToArray" deprecated="true" />
+    <rule ref="category/java/errorprone.xml/DontUseFloatTypeForLoopIndices" deprecated="true" />
+    <rule ref="category/java/errorprone.xml/JumbledIncrementer" deprecated="true" />
+    <rule ref="category/java/errorprone.xml/MisplacedNullCheck" deprecated="true" />
+    <rule ref="category/java/errorprone.xml/OverrideBothEqualsAndHashcode" deprecated="true" />
+    <rule ref="category/java/errorprone.xml/ReturnFromFinallyBlock" deprecated="true" />
+    <rule ref="category/java/errorprone.xml/UnconditionalIfStatement" deprecated="true" />
+
+    <rule ref="category/java/multithreading.xml/AvoidThreadGroup" deprecated="true" />
+    <rule ref="category/java/multithreading.xml/DontCallThreadRun" deprecated="true" />
+    <rule ref="category/java/multithreading.xml/DoubleCheckedLocking" deprecated="true" />
+
+    <rule ref="category/java/bestpractices.xml/AvoidUsingHardCodedIP" deprecated="true" />
+    <rule ref="category/java/bestpractices.xml/CheckResultSet" deprecated="true" />
+
+    <rule ref="category/java/codestyle.xml/ExtendsObject" deprecated="true" />
+    <rule ref="category/java/codestyle.xml/ForLoopShouldBeWhileLoop" deprecated="true" />
+
+    <rule ref="category/java/performance.xml/BigIntegerInstantiation" deprecated="true" />
+    <rule ref="category/java/performance.xml/BooleanInstantiation" deprecated="true" />
+
+    <rule ref="category/java/design.xml/CollapsibleIfStatements" deprecated="true" />
+    <rule ref="category/java/design.xml/SimplifiedTernary" deprecated="true" />
+
+</ruleset>
```

Comment:
Actually, the rule set needs to be written like this:
```
    <rule ref="category/java/bestpractices.xml" />
    <rule ref="category/java/codestyle.xml" />
    <rule ref="category/java/design.xml" />
    <rule ref="category/java/documentation.xml" />
    <rule ref="category/java/errorprone.xml" />
    <rule ref="category/java/multithreading.xml" />
    <rule ref="category/java/performance.xml" />
    <rule ref="category/java/security.xml" />
```
Then run and fix as many as possible. Some cannot be fixed or are not that critical (like upper case naming conventions), and those can be excluded.

Reply:
I'm not sure that you need to add "deprecated" to the end of each of these calls either.

Reply:
Yeah

Reply:
Writing the rules like that causes lots of warnings:
```
[WARNING] Unable to locate Source XRef to link to - DISABLED
[WARNING] Unable to locate Source XRef to link to - DISABLED
[WARNING] The rule AvoidDecimalLiteralsInBigDecimalConstructor is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule AvoidMultipleUnaryOperators is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule AvoidUsingOctalValues is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule BrokenNullCheck is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule CheckSkipResult is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule ClassCastExceptionWithToArray is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule DontUseFloatTypeForLoopIndices is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule JumbledIncrementer is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule MisplacedNullCheck is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule OverrideBothEqualsAndHashcode is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule ReturnFromFinallyBlock is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule UnconditionalIfStatement is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule AvoidDecimalLiteralsInBigDecimalConstructor is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule AvoidMultipleUnaryOperators is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule AvoidUsingOctalValues is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule BrokenNullCheck is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule CheckSkipResult is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule ClassCastExceptionWithToArray is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule DontUseFloatTypeForLoopIndices is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule JumbledIncrementer is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule MisplacedNullCheck is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule OverrideBothEqualsAndHashcode is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule ReturnFromFinallyBlock is referenced multiple times in "Basic". Only the last rule configuration is used.
[WARNING] The rule UnconditionalIfStatement is referenced multiple times in "Basic". Only the last rule configuration is used.
```

Please use the syntax I showed, it will eliminate all warnings. And put one rule per line, meaning that if you have an exception, write it on another line:
```
<rule ref="category/java/errorprone.xml">
     <exclude name="AvoidBranchingStatementAsLastInLoop" />
</rule>
```